src/persist-client/src/batch.rs

```diff
@@ -303,12 +250,120 @@ where
         }
         self.records.push(((&self.key_buf, &self.val_buf), t, d));
 
-        // TODO(aljoscha): As of now, this batch builder doesn't have constant
-        // memory usage but scales with the number of added records. Instead of
-        // keeping one records builder around, we should periodically finish and
-        // flush the updates to a blob, store the blob key in the builder, and
-        // reset the records builder.
+        // If we've filled up a chunk of ColumnarRecords, flush it out now to
+        // blob storage to keep our memory usage capped.
+        if let Some(parts) = self.records.take_filled() {
+            for part in parts {
+                let upper = Antichain::from_elem(self.max_ts.clone());
```

Comment:
not to self, I think this needs to increment max_ts somehow to be a proper upper

Reply:
I think yes, the `upper` of a batch must not be `<=` any of the timestamps in the batch.

Reply:
maybe parts don't have an upper? just spitballing...

Reply:
hmm they don't technically need the description at all. it's just a thing we have that we think might be useful for _hand-wave_ future debugging